packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+
+A schema could tell us that `node 2` is a "Dog" and that every dog has a string data payload - the dog's name - under `field key D`.
+In addition to knowing that the data in `node 7` is a string, and what that string means, we also know that it wouldn't make sense to put more than one node in that field - a dog can't have two names (not in this domain, anyway).
+
+If we have the guarantee that the data in the tree conforms to this "Dog" schema, then users can be confident about how to read and write the dog's name.
+We can also make the APIs better (for example, having a specific `setName(nameString)` function rather than a general `setField(fieldKey, nodes)` function).
+
+### Tree Schema Structure
+
+What does a tree schema actually look like?
+What aspects of the Intentional Tree Model introduced above might be constrained by a schema?
+The following list describes the building blocks for schema enforcement that are in use by SharedTree and have proven to be sufficient for typical SharedTree projects so far.
+
+1. **Nodes have a _type_.**
+   Knowing what a given node represents in an application domain is important to understanding the data in and around that node.
+   By assigning each node a unique type identifier (e.g. "Dog" or "com.microsoft.powerpoint.rectangle"), the data held by that node can be understood.
+
+2. **Nodes have expected fields.**
+   For a given type of node, a schema enumerates which fields are expected to be under that node.
+   For example, a node of type "Dog" is expected to have a field under the field key "name".
+
+   > Having an expected field means that the field is expected to be there, but the inverse is not true.
+     _Not_ having an expected field with key _K_ does not mean that there _can't_ be a field with key _K_.
+     Field _K_ might be in the tree, but it's invisible to a schema that doesn't know about it and that schema can (in theory) pretend it doesn't exist.
+
+   For each of these expected fields...
+
+   3. **Fields have a _kind_.**
+      The kind of a field determines its behavior (e.g. in the case of SharedTree, this might affect how concurrent edits to the field are merged).
+      One critical aspect of a field's kind is its _multiplicity_.
+      The multiplicity describes how many nodes a field is allowed to hold.
+      The Intentional tree model defines the following:
+
+      * **Required fields** must have exactly one node at all times.
+      * **Optional fields** can have either zero nodes or one node.
+      * **Sequence fields** can have any number of nodes (including zero).
+      * **Forbidden fields** must have exactly zero nodes (this is useful in some complicated scenarios but not something a typical user would desire).
```

Comment:
IMO, unless this is relevant (it's not mentioned anywhere else) I suggest either removing this or adding a concrete example of the use. It's not obvious why this is useful, so it's mostly confusing.